RavenDB-7574: WillCauseHardPageFault

aviviadi opened the pull request.

aviviadi committed:
RavenDB-7574: WillCauseHardPageFault

ayende commented on src/Sparrow/Platform/Posix/PosixMemoryQueryMethods.cs:
```diff
@@ -0,0 +1,29 @@
+            Debug.Assert(addr.ToInt64() % Syscall.PageSize == 0);
+
+            var vecSize = (length + Syscall.PageSize - 1) / Syscall.PageSize;
+            var vec = Marshal.AllocHGlobal((int)vecSize);
```
You are leaking this pointer.

aviviadi committed:
Reverse logic

ayende commented on src/Sparrow/Platform/Win32/Win32MemoryQueryMethods.cs:
```diff
@@ -0,0 +1,53 @@
+
+            }
+
+            for (int i = 0; i < pages; i++)
```
I think that the logic here needs to be reversed, given the new method name.

ayende commented on src/Sparrow/Platform/Posix/PosixMemoryQueryMethods.cs:
```diff
@@ -0,0 +1,29 @@
+            if (Syscall.mincore(addr.ToPointer(), new IntPtr(length), (char*)vec.ToPointer()) != 0)
+                throw new MemoryInfoException($"Failed to mincore addr: {addr.ToInt64()}, with length: {length}. Last Error = {Marshal.GetLastWin32Error()}");
+
+            for (var i = 0; i < vecSize; i++)
```
I think that the logic here needs to be reversed, given the new method name.

ayende commented on src/Sparrow/LowMemory/MemoryInformation.cs:
```diff
@@ -130,7 +130,7 @@ public static bool IsSwappingOnHddInsteadOfSsd()
         }
 
-        public static bool WillCauseHardPageFault(IntPtr addr, long length) => PlatformDetails.RunningOnPosix ? PosixMemoryQueryMethods.WillCauseHardPageFault(addr, length) : Win32MemoryQueryMethods.WillCauseHardPageFault(addr, length);
+        public static bool WillCauseHardPageFault(IntPtr addr, long length) => PlatformDetails.RunningOnPosix ? !PosixMemoryQueryMethods.WillCauseHardPageFault(addr, length) : !Win32MemoryQueryMethods.WillCauseHardPageFault(addr, length);
```
No, this is _REALLY_ confusing. We'll typically look at the impl, not the wrapper

aviviadi committed:
RavenDB-7574:
* Reverse logic
* Use byte* instead of IntPtr in calling method
* Use stackalloc for small sizes (only in posix, as in windows even for one page we need at least 2 times size of long)
MacOSX Remark : syspage is per platform so it take into consideration macosx, however mincore need to be tested separately on mac. This code is not in use yet anywhere..

ayende commented on src/Sparrow/Platform/Posix/PosixMemoryQueryMethods.cs:
```diff
@@ -7,23 +7,45 @@ namespace Sparrow.Platform.Posix
-            var vec = Marshal.AllocHGlobal((int)vecSize);
-            if (Syscall.mincore(addr.ToPointer(), new IntPtr(length), (char*)vec.ToPointer()) != 0)
-                throw new MemoryInfoException($"Failed to mincore addr: {addr.ToInt64()}, with length: {length}. Last Error = {Marshal.GetLastWin32Error()}");
+            var vecSize = (int)((length + Syscall.PageSize - 1) / Syscall.PageSize);
```
Overflow?

ayende commented on src/Sparrow/Platform/Win32/Win32MemoryQueryMethods.cs:
```diff
@@ -25,19 +25,19 @@ struct PPSAPI_WORKING_SET_EX_INFORMATION
             {
                 if (QueryWorkingSetEx(GetCurrentProcess(), (byte*)pWsInfo, (uint)(sizeof(PPSAPI_WORKING_SET_EX_INFORMATION) * pages)) == false)
-                    throw new MemoryInfoException($"Failed to QueryWorkingSetEx addr: {addr.ToInt64()}, with length: {length}. processId = {GetCurrentProcess()}");
+                    throw new MemoryInfoException($"Failed to QueryWorkingSetEx addr: {new IntPtr(addr).ToInt64()}, with length: {length}. processId = {GetCurrentProcess()}");
```
The logic here is still wrong.

aviviadi committed:
RavenDB-7574: revert logic also in win, beware not to exceed legnth > int.MaxValue, and stackalloc small sizes in win too

ayende commented on src/Sparrow/Platform/Win32/Win32MemoryQueryMethods.cs:
```diff
@@ -27,27 +27,38 @@ struct PPSAPI_WORKING_SET_EX_INFORMATION
-                if (QueryWorkingSetEx(GetCurrentProcess(), (byte*)pWsInfo, (uint)(sizeof(PPSAPI_WORKING_SET_EX_INFORMATION) * pages)) == false)
-                    throw new MemoryInfoException($"Failed to QueryWorkingSetEx addr: {new IntPtr(addr).ToInt64()}, with length: {length}. processId = {GetCurrentProcess()}");
-
+                var wsInfo = Marshal.AllocHGlobal((int)(sizeof(PPSAPI_WORKING_SET_EX_INFORMATION)*pages));
```
You are leaking this memory!

aviviadi replied:
https://github.com/ravendb/ravendb/pull/3032/commits/e3c3071945eea947886110bdbd7ec3b8f51337eb

aviviadi committed:
Avoid mem leak
